proposals/0001-metadata-providers.md

```diff
@@ -0,0 +1,489 @@
+# Metadata Providers
+
+Authors: [Moritz Lang](https://github.com/slashmo), [Konrad 'ktoso' Malawski](https://github.com/ktoso)
+
+## Introduction
+
+While global metadata attributes may be manually set on a `LogHandler` level, there's currently no way of reliably providing contextual, automatically propagated, metadata when logging with swift-log.
+
+## Motivation
+
+To benefit from tools such as [Distributed Tracing](https://github.com/apple/swift-distributed-tracing) it is necessary for libraries to make use of the trace information.
+
+Most notably, loggers should participate in tracing by including some trace metadata (such as e.g. a `trace-id`) when logging, as it transparently enables developers to benefit from log correlation using those IDs.
+
+Today, the only supported way of providing such metadata is to pass them along to each log call explicitly:
+
+```swift
+logger.info("first this ...", metadata: ["trace-id": MyTracingLibrary.currentTraceID])
+logger.info("... now this", metadata: ["trace-id": MyTracingLibrary.currentTraceID])
+```
+
+This comes with a couple of downsides:
```

Comment:
Just as a general note I think this is not the only supported way. The other way, which is also the one that I see more often, is copying the `Logger` once you decoded the `trace-id` and from that point on pass that `Logger` forward.

Reply:
Good point, maybe the example lacks a bit. In practice it wouldn't just be about passing around a trace ID, but also a span ID. A span ID can't be defined wherever you create the `Logger` which you pass around, because it's changing during the lifetime of a single request.